ega-data-api-dataedge/src/main/java/htsjdk/samtools/seekablestream/EgaSeekableCachedResStream.java

```diff
@@ -0,0 +1,310 @@
+    public EgaSeekableCachedResStream setExtension(String hack_extension) {
+        this.hack_extension = hack_extension;
+        return this;
+    }
+
+    // ------------------------------------------------------------------------- Cache Population
+    // ------------------------------------------------------------------------- Guava Cache
+
+    // separate from stream reading/position
+    private byte[] get(int page_number) throws ExecutionException {
+        int maxPage = (int) (this.contentLength / PAGE_SIZE + 1); // Don'd read past end of stream
+
+        //int firstPage = page_number>9?page_number-10:0; // Get prior cache page, just in case
+        int firstPage = page_number > 0 ? page_number - 1 : 0; // Get prior cache page, just in case
+        int lastPage = (page_number + NUM_PAGES - 1) > maxPage ? maxPage : (page_number + NUM_PAGES - 1);
+
+        for (int i = firstPage; i < lastPage; i++) {
+            final int page_i = i;
+            new Thread(() -> {
+                try {
+                    this.cache.get(page_i);
+                } catch (ExecutionException e) {
+                }
+            }).start();
+        }
+
+        //    byte[] get = this.cache.get(page_number);
+        //    while (get[0]==0 && get[1]==0) {
+        //        try {Thread.sleep(50);} catch (InterruptedException ex) {}
+        //        get = this.cache.get(page_number);
+        //    }
+
+        //	return get;
+
+        return this.cache.get(page_number);
+    }
+
+    private byte[] populateCache(int page_number) {
+        // Last Page Handling
+        int maxPage = (int) (this.contentLength / PAGE_SIZE + 1) - 1; // Don'd read past end of stream
+        if (page_number > maxPage)
+            return new byte[]{};
+
+        long offset = (long) page_number * (long) PAGE_SIZE;
+        final int bytesToRead = (int) ((page_number == maxPage) ? (this.contentLength - offset) : PAGE_SIZE);
+
+        // Prepare buffer to read from file
+        byte[] bytesRead = new byte[bytesToRead];
+        int totalBytesRead = 0;
+
+        synchronized (this) {
+            try {
+                String byteRange = "bytes=" + offset + "-" + (offset + bytesToRead);
+                String url = this.url.toString() + "?startCoordinate=" + offset +
+                        "&endCoordinate=" + (offset + bytesToRead) +
+                        "&destinationFormat=" + "Plain";
+
+                Request datasetRequest = new Request.Builder()
+                        .url(url)
+                        .build();
+
+                // Execute the request and retrieve the response.
+                okhttp3.Response response = client.newCall(datasetRequest).execute();
+                ResponseBody body = response.body();
+
+                InputStream byteStream = response.body().byteStream();
+                byte[] buff = new byte[8000];
+                ByteArrayOutputStream bao = new ByteArrayOutputStream();
```

Comment:
Can you please use try with resource for all Input/Output streams.

Reply:
The documentation says "Closing a ByteArrayOutputStream has no effect."

Documentation: https://docs.oracle.com/javase/7/docs/api/java/io/ByteArrayOutputStream.html